src/Corax/Queries/SortingMatches/SortingMatch.cs

```diff
@@ -0,0 +1,385 @@
+        var sortedIdsScope = allocator.Allocate( sizeof(long) * SortBatchSize, out bs);
+        Span<long> sortedIdBuffer = new(bs.Ptr, SortBatchSize);
+
+        SortedIndexReader<TDirection> reader = GetReader(ref match);
+        
+        while (match._results.Count < maxResults)
+        {
+            var read = reader.Read(sortedIdBuffer);
+            if (read == 0)
+                break;
+            var sortedIds = sortedIdBuffer[..read];
+            var indexes = indexesBuffer[..read];
+            // we effectively permute the indexes as well as the sortedIds to get a sorted list to compare
+            // with the allMatches
+            InitializeIndexesTopHalf(indexes);
+            sortedIds.Sort(indexes);
+            InitializeIndexesBottomHalf(indexes);
+            read = SortHelper.FindMatches(indexes, sortedIds, allMatches);
+            indexes = indexes[..read];
+            indexes.Sort();
+            // now get the *actual* matches in their sorted order
+            for (int i = 0; i < indexes.Length; i++)
+            {
+                match._results.Add(sortedIds[(int)indexes[i]]);
+            }
+        }
+
+        sortedIdsScope.Dispose();
+        indexesScope.Dispose();
+        
+        
+        SortedIndexReader<TDirection> GetReader(ref SortingMatch<TInner> match)
+        {
+            if (typeof(TDirection) == typeof(CompactTree.ForwardIterator) ||
+                typeof(TDirection) == typeof(CompactTree.BackwardIterator))
+            {
+                var termsTree = match._searcher.GetTermsFor(entryCmp.GetSortFieldName(ref match));
+                return new SortedIndexReader<TDirection>(llt, match._searcher, termsTree.Iterate<TDirection>());
+            }
+
+            if (typeof(TDirection) == typeof(Lookup<long>.ForwardIterator) ||
+                typeof(TDirection) == typeof(Lookup<long>.BackwardIterator))
+            {
+                var termsTree = match._searcher.GetLongTermsFor(entryCmp.GetSortFieldName(ref match));
+                return new SortedIndexReader<TDirection>(llt, match._searcher, termsTree.Iterate<TDirection>());
+            }
+
+            if (typeof(TDirection) == typeof(Lookup<double>.ForwardIterator) ||
+                typeof(TDirection) == typeof(Lookup<double>.BackwardIterator))
+            {
+                var termsTree = match._searcher.GetDoubleTermsFor(entryCmp.GetSortFieldName(ref match));
+                return new SortedIndexReader<TDirection>(llt, match._searcher, termsTree.Iterate<TDirection>());
+            }
+
+            throw new NotSupportedException(typeof(TDirection).FullName);
+        }
+        
+
+    }
+
+    private static void InitializeIndexesTopHalf(Span<long> span)
+    {
+        for (int i = 0; i < span.Length; i++)
+            span[i] = (long)i << 32;
+    }
+    
+    private static void InitializeIndexesBottomHalf(Span<long> span)
+    {
+        for (int i = 0; i < span.Length; i++)
+            span[i] |= (uint)i;
```

Comment:
`InitializeIndexesTopHalf` ensures that the bottom 32 bits will become `0`, then we sort based on that and then we ensure that we can number from `0..n` the elements . While it would be the same to sum the `i`, since all bits are `0` (guaranteed) the fastest operation is to `or` the low part of the register

Reply:
Thanks. Laving the comment as unresolved on purpose so it will be visible in the PR